enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,420 @@
+will be that our existing test coverage does not regress when upgrading to
+external CCM.
+
+We will ensure full coverage of edge cases which are impractical to test on real
+hardware through unit testing.
+
+#### Existing OpenStack test coverage
+
+QE already regularly runs a full test suite against a matrix of the following
+test configurations:
+
+- Networking:
+  - OpenshiftSDN
+  - OVNKubernetes
+  - Kuryr
+- Installation:
+  - IPI
+  - IPI with Proxy
+  - UPI
+- OpenStack version:
+  - OSP 16.1
+  - OSP 16.2
+
+Specifically, they test:
+- conformance/parallel
+- cinder-csi
+- manilla
+- openstack-test
+
+The test configurations already cover the full range of documented
+configurations changes, as can be seen in the [test code][cloud-cm-gerrit].
+
+[cloud-cm-gerrit]: https://code.engineering.redhat.com/gerrit/gitweb?p=openshift-ir-plugin.git;a=blob;f=roles/tests/tasks/change_cloud_cm.yml;h=33ee741b1ef9c45c0417b911d234b4d3d83e68e2;hb=HEAD
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+N/A. This feature is part of CCM, which is already tech preview. It will become
+the default for OpenStack-based clusters in 4.11.
+
+#### Tech Preview -> GA
+
+QE's test coverage of TechPreview is limited, but it has been sufficient to
+discover this omission. When we enable CCM by default in 4.11 we will be covered
+by QE's full test matrix detailed above. This feature will graduate to GA if
+these tests are passing.
+
+Additionally:
+- We will land the feature as early as possible in the 4.11 development cycle
+- User facing documentation is updated to highlight configuration options which
+  are deprecated between 4.10 and 4.11
+- User facing documentation is updated to show the new configuration
+
+#### Removing a deprecated feature
+
+N/A. This is a new feature.
+
+### Upgrade / Downgrade Strategy
+
+N/A: Downgrade from CCM is not supported
+
+### Version Skew Strategy
+
+Version skew is handled by maintaining the legacy configuration while writing
+the new configuration to a new configmap. This strategy does not require any
+changes to the existing external CCM upgrade workflow: legacy components
+continue to read the legacy configuration until they are restarted, and the new
+component can read the new configuration at any time.
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+In general, failure of CCO will not have a critical impact on the cluster unless
+the user has also manually deleted the configmaps it creates in
+`openshift-config-managed`.
+
+A failure of CCO will cause updates to user-managed config to not be reflected
+in the cluster's deployed configuration. However, the cluster will continue to
+use the previous configuration until the failure of CCO is resolved.
+
+#### Support Procedures
+
+If CCO has failed it will be evident because it will be marked as degraded. If
+running 4.10 this will prevent an upgrade to 4.11. If running 4.11, a valid
```

Comment:
If a 4.10 operator wants to block updates to 4.11, it should go `Upgradeable=False`, not `Degraded=True`.  `Degraded=True` is "there are quality-of-service impacts right now, and we need a human admin to help resolve soon".  `Upgradeable=False` is "might be fine now, but we are not in a good place vs. the next minor, and will need a human admin to help before you want to go to the next minor".

Reply:
Done